### PR TITLE
Adding POJOs mapping AWS Price List API JSON object

### PR DESCRIPTION
**Pull request by LSZ-PL**
Adding AWS pricing supplier with Pricing Canonical Data Model
Adding multi cloud pricing service and integrating into existing multi cloud service plus configuring Guice modules plus pom dependencies

### CODE REVIEW

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+            for(AWSProductAndTerms awsProductAndTerms : awsPricingPack.getAwsProductAndTerms()) {
+                if(!productFamiliesFilter.contains(awsProductAndTerms.product.productFamily)
+                        || !tenanciesFilter.contains(awsProductAndTerms.product.attributes.get("tenancy"))
+                        || !capacityStatusesFilter.contains(awsProductAndTerms.product.attributes.get("capacitystatus"))
+                        || !awsProductAndTerms.product.attributes.get("operation").startsWith(operationsPatternFilter)) {
+                    continue;
+                }
+
+                OperatingSystemArchitecture operatingSystemArchitecture;
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: Choosing OperatingSystemArchitecture could be externalized to separate method.

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                        operatingSystemArchitecture = OperatingSystemArchitecture.I368;
+                        break;
+                    }
+                    default: {
+                        operatingSystemArchitecture = OperatingSystemArchitecture.UNKNOWN;
+                        break;
+                    }
+                }
+
+                OperatingSystemFamily operatingSystemFamily;
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: OperatingSystemFamily could be externalized to separate method.

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                                pricingDimensions.add(new PricingDimensions(price.getValue(), entry3.getValue().getUnit(), entry3.getValue().getDescription(), entry3.getValue().getBeginRange(), entry3.getValue().getEndRange()));
+                            }
+                            pricingOnDemandTerms.add(new PricingTerms(pricingDimensions, entry.getValue().termAttributes.get("LeaseContractLength"), entry.getValue().termAttributes.get("OfferingClass"), entry.getValue().termAttributes.get("PurchaseOption")));
+                        }
+                    }
+                    currency = Optional.ofNullable(awsProductAndTerms.terms.get("OnDemand").entrySet().iterator().next().getValue().getAwsDimensionRates().entrySet().iterator().next().getValue().entrySet().iterator().next().getValue().getPricePerUnit().entrySet().iterator().next().getKey());
+                }
+
+                // currently only OnDemand offers are being presented
+                /*if (awsProductAndTerms.terms != null && awsProductAndTerms.terms.get("Reserved") != null) {
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: Comment to remove?

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                    if(!currency.isPresent()) {
+                        currency = Optional.ofNullable(awsProductAndTerms.terms.get("Reserved").entrySet().iterator().next().getValue().getAwsDimensionRates().entrySet().iterator().next().getValue().entrySet().iterator().next().getValue().getPricePerUnit().entrySet().iterator().next().getKey());
+                    }
+                }*/
+
+                Pricing pricing = PricingBuilder.newBuilder()
+                        .id(awsProductAndTerms.product.sku)
+                        .providerId(awsProductAndTerms.product.sku)
+                        .name(awsProductAndTerms.product.attributes.get("usagetype"))
+                        .location(null)
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: Probably not needed - null is default value.

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                            for (Map.Entry<String, AWSBaseTerm> entry3 : entry2.getValue().entrySet()) {
+                                Map.Entry<String, BigDecimal> price = entry3.getValue().getPricePerUnit().entrySet().iterator().next();
+                                pricingDimensions.add(new PricingDimensions(price.getValue(), entry3.getValue().getUnit(), entry3.getValue().getDescription(), entry3.getValue().getBeginRange(), entry3.getValue().getEndRange()));
+                            }
+                            pricingOnDemandTerms.add(new PricingTerms(pricingDimensions, entry.getValue().termAttributes.get("LeaseContractLength"), entry.getValue().termAttributes.get("OfferingClass"), entry.getValue().termAttributes.get("PurchaseOption")));
+                        }
+                    }
+                    currency = Optional.ofNullable(awsProductAndTerms.terms.get("OnDemand").entrySet().iterator().next().getValue().getAwsDimensionRates().entrySet().iterator().next().getValue().entrySet().iterator().next().getValue().getPricePerUnit().entrySet().iterator().next().getKey());
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: This oneLine'r is a bit too long... ;)

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+                        .architecture(operatingSystemArchitecture)
+                        .family(operatingSystemFamily)
+                        .version(OperatingSystemVersions.unknown()); // AWS Pricing API exposes no information about version
+
+                List<PricingTerms> pricingOnDemandTerms = new ArrayList<>();
+                List<PricingTerms> pricingReservedTerms = new ArrayList<>();
+                Optional<String> currency = Optional.empty();
+
+                if(awsProductAndTerms.terms != null && awsProductAndTerms.terms.get("OnDemand") != null) {
+                    for (Map.Entry<String, AWSPriceDimensionsRates> entry : awsProductAndTerms.terms.get("OnDemand").entrySet()) {
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: entry, entry2, entry3...

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                        .preInstalledSw(awsProductAndTerms.product.attributes.get("preInstalledSw"))
+                        .productFamily(awsProductAndTerms.product.productFamily)
+                        .operation(awsProductAndTerms.product.attributes.get("operation"))
+                        .build();
+                pricings.add(pricing);
+            }
+
+            nextToken = awsPricingPack.getNextToken();
+        } while (nextToken != null);
+        LOGGER.info(String.format("AWS products (with terms) processed (= size of returned Set<Pricing>): %d", pricings.size()));
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: Logger has support for placeholders, so its the same:
LOGGER.info("AWS products (with terms) processed (= size of returned Set<Pricing>): {}}", pricings.size());

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+                    default: {
+                        operatingSystemFamily = OperatingSystemFamily.UNKNOWN;
+                        break;
+                    }
+                }
+
+                OperatingSystemBuilder operatingSystemBuilder = OperatingSystemBuilder.newBuilder()
+                        .architecture(operatingSystemArchitecture)
+                        .family(operatingSystemFamily)
+                        .version(OperatingSystemVersions.unknown()); // AWS Pricing API exposes no information about version
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: .build() could be invoked at the end of this statement.

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSProduct.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package de.uniulm.omi.cloudiator.sword.multicloud.pricing.aws;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import java.util.Map;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AWSProduct {
+    public String sku;
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSProduct.java]: all field public?

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSProductAndTerms.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package de.uniulm.omi.cloudiator.sword.multicloud.pricing.aws;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AWSProductAndTerms {
+    public AWSProduct product;
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSProductAndTerms.java]: public?

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSTermRates.java [hunk 11]
```diff
@@ -0,0 +1,8 @@
+package de.uniulm.omi.cloudiator.sword.multicloud.pricing.aws;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import java.util.HashMap;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AWSTermRates extends HashMap<String, AWSPriceDimensionsRates> {}
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSTermRates.java]: Move closing bracket to new line.

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSTerms.java [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package de.uniulm.omi.cloudiator.sword.multicloud.pricing.aws;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import java.util.HashMap;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AWSTerms extends HashMap<String, AWSTermRates> {}
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSTerms.java]: Move closing brackets to new line.

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/service/MultiCloudPricingService.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class MultiCloudPricingService implements PricingService {
+    private final CloudRegistry cloudRegistry;
+    @Inject
+    private PricingSupplierFactory pricingSupplierFactory;
+
+    @Inject
+    public MultiCloudPricingService(CloudRegistry cloudRegistry) {
+        checkNotNull(cloudRegistry, "cloudRegistry is null");
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/service/MultiCloudPricingService.java]: Checking if not null and variable assignment could be done in one line.

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/service/MultiCloudPricingService.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        this.cloudRegistry = cloudRegistry;
+    }
+
+    @Override
+    public Iterable<Pricing> listPricing() {
+        final ImmutableSet.Builder<Pricing> builder = ImmutableSet.builder();
```
Comment (pszkup7bulls) [inline multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/service/MultiCloudPricingService.java]: Body of this method could be replaced by:

`        final ImmutableSet.Builder<Pricing> builder = ImmutableSet.builder();
        cloudRegistry.
                list()
                .stream()
                .filter(cloud -> cloud.api().providerName().equals("aws-ec2"))
                .findFirst()
                .ifPresent(cloud -> builder.addAll(pricingSupplierFactory.createAWSPricingSupplier(cloud.credential()).get()));
        return builder.build();`

but I am sure that it still could be done in more functional way...

File: core/src/main/java/de/uniulm/omi/cloudiator/sword/domain/PricingImpl.java [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+        return super.toStringHelper()
+                .add("instanceName", this.instanceName)
+                .add("cloudServiceProviderName", this.cloudServiceProviderName)
+                .add("locationProviderId", this.locationProviderId)
+                .add("api", this.api.providerName())
+                .add("currency", this.currency)
+                .add("tenancy", this.tenancy)
+                .add("licenseModel", this.licenseModel)
+                .add("operatingSystem", this.operatingSystem)
+                .add("pricingOnDemandTerms.size", (this.pricingOnDemandTerms != null ? this.pricingOnDemandTerms.size() : -1))
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply core/src/main/java/de/uniulm/omi/cloudiator/sword/domain/PricingImpl.java]: -1 means that the list in question was simply null

File: core/src/main/java/de/uniulm/omi/cloudiator/sword/domain/PricingImpl.java [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+                .add("pricingReservedTerms.size", (this.pricingReservedTerms != null ? this.pricingReservedTerms.size() : -1))
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply core/src/main/java/de/uniulm/omi/cloudiator/sword/domain/PricingImpl.java]: -1 means that the list in question was simply null

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/MultiCloudServiceImpl.java [hunk 17]
```diff
@@ -31,13 +32,16 @@
 
   private final CloudRegistry cloudRegistry;
   private final ComputeService computeService;
+  private final PricingService pricingService;
 
   @Inject
-  public MultiCloudServiceImpl(CloudRegistry cloudRegistry, ComputeService computeService) {
+  public MultiCloudServiceImpl(CloudRegistry cloudRegistry, ComputeService computeService, PricingService pricingService) {
     checkNotNull(cloudRegistry, "cloudRegistry is null");
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/MultiCloudServiceImpl.java]: done, now it's beautiful - I'm serious

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/PricingSupplierFactory.java [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+
+import com.google.common.base.Supplier;
+import com.google.inject.name.Named;
+import de.uniulm.omi.cloudiator.sword.domain.CloudCredential;
+import de.uniulm.omi.cloudiator.sword.domain.Pricing;
+
+import java.util.Set;
+
+public interface PricingSupplierFactory {
+    @Named("aws") Supplier<Set<Pricing>> createAWSPricingSupplier(CloudCredential cloudCredential);
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/PricingSupplierFactory.java]: done

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSBaseTerm.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+import java.math.BigDecimal;
+import java.util.HashMap;
+import java.util.Map;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AWSBaseTerm {
+    @JsonIgnore
+    private Map<String, Object> additionalProperties = new HashMap<String, Object>();
+
+    @JsonProperty("unit")
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSBaseTerm.java]: removed unnecessary @JsonProperty annotations

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSBaseTerm.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+        return this.additionalProperties;
+    }
+
+    @JsonAnySetter
+    public void setAdditionalProperty(String name, Object value) {
+        this.additionalProperties.put(name, value);
+    }
+
+    @JsonCreator
+    public AWSBaseTerm(@JsonProperty("unit") String unit, @JsonProperty("description") String description, @JsonProperty("beginRange") String beginRange,
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSBaseTerm.java]: yes, explained here:
https://fasterxml.github.io/jackson-annotations/javadoc/2.7/com/fasterxml/jackson/annotation/JsonCreator.html

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+                Pricing pricing = PricingBuilder.newBuilder()
+                        .id(awsProductAndTerms.product.sku)
+                        .providerId(awsProductAndTerms.product.sku)
+                        .name(awsProductAndTerms.product.attributes.get("usagetype"))
+                        .location(null)
+                        .locationProviderId(awsPricingLocationConverter.apply(awsProductAndTerms.product.attributes.get("location")))
+                        .instanceName(awsProductAndTerms.product.attributes.get("instanceType"))
+                        .cloudServiceProviderName("aws")
+                        .api(ApiBuilder.newBuilder().providerName("aws-ec2").build())
+                        .currency(currency.isPresent() ? currency.get() : "N/A")
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: done - now it's slick

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+public class AWSPricingSupplier implements Supplier<Set<Pricing>> {
+
+    private final AWSCredentialsProvider credentialsProvider;
+    private final AWSPricing client;
+    private final AWSPricingLocationConverter awsPricingLocationConverter = new AWSPricingLocationConverter();
+    private static final Logger LOGGER = LoggerFactory
+            .getLogger(AWSPricingSupplier.class);
+    private static final List<String> productFamiliesFilter = Arrays.asList("Compute Instance", "Compute Instance (bare metal)");
+    private static final List<String> tenanciesFilter = Arrays.asList("Shared", "Dedicated");
+    private static final List<String> capacityStatusesFilter = Arrays.asList("Used");
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: done

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+        credentialsProvider = new AWSStaticCredentialsProvider(awsCredentials);
+        client = AWSPricingClientBuilder.standard().withRegion("us-east-1").withCredentials(credentialsProvider).build();
+        LOGGER.info("AWSPricing client (with region us-east-1) created");
+    }
+
+    private AWSPricingPack getRawAWSPricingPaginated(String nextToken) {
+        GetProductsRequest getProductsRequest = new GetProductsRequest().withServiceCode("AmazonEC2");
+        GetProductsResult getProductsResult;
+
+        if (nextToken == null || nextToken.isEmpty()){
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: done - now it's slick

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+
+        Set<AWSProductAndTerms> awsProductAndTermsSet = new HashSet<>(100);
+
+        try {
+            for(int i=0; i < getProductsResult.getPriceList().size(); i++) {
+                awsProductAndTermsSet.add(mapper.readValue(getProductsResult.getPriceList().get(i), AWSProductAndTerms.class));
+            }
+        } catch (Exception e) {
+            LOGGER.error("Exception caught while retrieving AWS Price List", e);
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: good catch ;)

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+        AWSPricingPack awsPricingPack = null;
+        Set<Pricing> pricings = new HashSet<>();
+
+        int awsPricingPackCounter = 0;
+        LOGGER.info("Starting fetching AWS Pricing Information");
+
+        do {
+            awsPricingPack = getRawAWSPricingPaginated(nextToken);
+
+            /*if(!awsPricingPack.getAwsProductAndTerms().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: removed

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java (hunk 1 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: leaving as is

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java (hunk 2 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: leaving as is

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java (hunk 3 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: I would prefer to leave it, as Reserved instances are planned to be included in near future

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java (hunk 4 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: this method expects `Location location` which I'm not passing in, so null is required

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java (hunk 5 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: I broke it into several lines of code

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java (hunk 6 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: changed to be more specific

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java (hunk 7 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: done

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java (hunk 8 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSPricingSupplier.java]: leaving as is

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSProduct.java (hunk 9 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSProduct.java]: I'm gonna leave it as is - it's just a pojo

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSProductAndTerms.java (hunk 10 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSProductAndTerms.java]: I'm gonna leave it as is - it's just a pojo

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSTermRates.java (hunk 11 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSTermRates.java]: done

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSTerms.java (hunk 12 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/pricing/aws/AWSTerms.java]: done

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/service/MultiCloudPricingService.java (hunk 13 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/service/MultiCloudPricingService.java]: done

File: multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/service/MultiCloudPricingService.java (hunk 14 above)
Comment (LSZ-PL) [reply multicloud/src/main/java/de/uniulm/omi/cloudiator/sword/multicloud/service/MultiCloudPricingService.java]: used suggested code

**Comment by LSZ-PL** [comment]
Hello Daniel,
please review my PR containing pricing functionality (pricing discovery + persistence + querying) and processing of first cloud service provider - AWS.

Thanks,
ŁSz

**Comment by dbaur** [comment]
@LSZ-PL please check the merge conflicts

**Comment by dbaur** [comment]
@LSZ-PL I solved the merge conflict, however this PR does not compile...

**Pull request merged**